### DISCUSSION TITLE
New Tutorial: Multi-Agent Coordination Tutorial Now Available!

### DISCUSSION THREAD

**Discussion by aditjha** [prl-mushr/mushr#44] [General]
There is a new tutorial posted on the MuSHR website! This tutorial is Multi-Agent Coordination Planner for Multi-Goal Tasks. We hope this is useful for many and let us know if there are any questions. Below is a link to the tutorial in case. 

https://mushr.io/tutorials/coordination/

**Comment by Lilly-yang** [comment]
Hi,

First of all, thank you for writing this detailed tutorial. I learned a lot from it.

It just I encountered an error when running the last step `init_planner.launch`. Here are screenshots.
`Mushr_coordination.launch`
![Screenshot from 2022-04-07 18-21-48](https://user-images.githubusercontent.com/39558023/162252393-2d956c2b-5b23-43e0-9133-b941eb59ec52.png)

`init_planner.launch`
![Screenshot from 2022-04-07 18-22-20](https://user-images.githubusercontent.com/39558023/162252618-a8adc738-b2d9-4988-85c6-36c3cb3691ea.png)

`rviz`
![Screenshot from 2022-04-07 18-22-56](https://user-images.githubusercontent.com/39558023/162252772-1affd3a6-aa27-4001-89a0-9a7e8c08413d.png)

Could you please help me to find out why `init_planner` closes after publishing all these plans for these 4 cars (of course, the cars in rviz didn't move either)?

I also found that after launching the [`multi_teleop.launch`](https://mushr.io/tutorials/coordination_planner/multi_teleop.launch) file provided in the tutorial, I can only see one car in rviz. I found another file with the same name in the [nhttc_ros](https://github.com/prl-mushr/nhttc_ros/blob/main/launch/multi_teleop.launch) repository which allows four carts to be displayed in rviz successfully. The latter launchs one more `nhttc_pose_init` node than the former, and it seems that without `nhttc_pose_init` the carts cannot be displayed in RIVZ (but I don't know why).
![Screenshot from 2022-04-07 18-55-26](https://user-images.githubusercontent.com/39558023/162257295-3db083cf-8ab4-4146-b543-99a5538affca.png)

![Screenshot from 2022-04-07 18-55-45](https://user-images.githubusercontent.com/39558023/162257330-30c28c52-c080-4c1e-8925-a9af3d128760.png)

Thank you!

**Comment by sidtalia** [comment]
Hi @Lilly-yang!
1) Init-planner closes after publishing the plans because once the plan has been published, it (the planner) is no longer required. This is because the system (at the moment) does not consider re-planning.

2) The purpose of nhttc_pose_init is to initialize the poses for the cars. The reason why you may be seeing only one car in rviz when you launch multi_teleop.launch from mushr_sim is that all of them are being "plopped" on the same coordinate (origin), thus it looks like its just one car. In nhttc_pose_init, the cars are plopped at different locations, which is why you see them separately.

3) The cars don't move after the init-planner closes because the init-planner node is only responsible for publishing plans. The multi_teleop.launch file being referred to in this tutorial belongs to the mushr_sim repository, which means it does not contain the local-navigation system provided by nhttc_ros and is only used to bring the cars into the "arena". I believe the tutorial needs to be updated to point to the correct launch file for multi-agent navigation. In order for the cars to move, you need to launch the multi_teleop.launch in nhttc_ros. If I remember correctly, you should launch the nhttc_ros node before launching the planner. This is because the nhttc_pose_init 
simply starts the cars off at some pre-determined location, whereas the init_planner node sets the cars on the appropriate start points.

Regards,
Sidharth T.

**Reply by Lilly-yang** [reply]
Hi Sidharth,

Thanks for your reply.

The `multi_teleop.launch` file I launched is from `nhttc_ros`. But it still can't move these cars. All `controller` nodes are dead after publishing plans. The steps I have done are as follows:
```
roslaunch mushr_coordination mushr_coordination.launch
roslaunch nhttc_ros multi_teleop.launch
rviz
roslaunch mushr_coordination init_planner.launch
```
(cars_file:=4cars.yaml tasks_file:=4cars4tasks5x3.yaml)
![Screenshot from 2022-04-08 16-59-14](https://user-images.githubusercontent.com/39558023/162467938-3211ecee-17b8-4a97-a734-ab216a3852e7.png)
![Screenshot from 2022-04-08 16-53-01](https://user-images.githubusercontent.com/39558023/162468003-4290f4f7-1a6d-4f9b-a9ac-347af82a56f2.png)

**Comment by sidtalia** [comment]
Hi @Lilly-yang,

Could you try the following:
```
roscore
(new tab)
roslaunch nhttc_ros multi_teleop.launch
wait for a few minutes till you see this output:
[ INFO] [1621678112.116392844]: node started
[ INFO] [1621678112.116614789]: carrot_goal_ratio: 1.000000
[ INFO] [1621678112.116632550]: max_ttc: 6.000000
[ INFO] [1621678112.116642780]: solver_time: 20
[ INFO] [1621678112.116653036]: obey_time:1
.......
(it would look similar to what has been shown above, though it may contain more fields. Note that this can take a while (~1 minute).
(new tab)
rosrun rviz rviz  (or just rviz)
roslaunch mushr_coordination mushr_coordination.launch
(wait for it to initialize, the new tab)
roslaunch mushr_coordination init_planner.launch
```
The nhttc ros nodes tend to die out when launched around the same time as the rviz, so try waiting for the node to start (the info messages will indicate that it has started).

Regards,
Sidharth T.

**Reply by Lilly-yang** [reply]
Hi Sidharth,

I just tried it. I waited for about 5 minutes and didn't see this
```
[ INFO] [1621678112.116392844]: node started
[ INFO] [1621678112.116614789]: carrot_goal_ratio: 1.000000
[ INFO] [1621678112.116632550]: max_ttc: 6.000000
[ INFO] [1621678112.116642780]: solver_time: 20
[ INFO] [1621678112.116653036]: obey_time:1
```

it stopped at
```
[initializer-51] process has finished cleanly
log file: /home/li/.ros/log/8774fc88-b751-11ec-82f6-0028f8a19a14/initializer-51*.log
```
Once I launch` init_planner.launch`, it reports the error that controllers' nodes are dead.
![Screenshot from 2022-04-08 17-57-51](https://user-images.githubusercontent.com/39558023/162479086-366967a3-f984-4f2b-84a2-a72e84247fb0.png)

BTW. I noticed that there might be a '2' missing [here](https://github.com/prl-mushr/nhttc_ros/blob/183f0610bf35d5bd4c88e8b8d6ccb38fd89cac14/launch/multi_teleop.launch#L40).

**Reply by sidtalia** [reply]
hi! good catch, does fixing that fix the problem? I'm not sure if it would, but it may be worth a shot.
Here's one more thing you should try:
inside the nhttc_ros multi_teleop.launch file, change [this](https://github.com/prl-mushr/nhttc_ros/blob/183f0610bf35d5bd4c88e8b8d6ccb38fd89cac14/launch/multi_teleop.launch#L17) line to the following:
```
<include file="$(find mushr_sim)/launch/map_server.launch" />
```

**Reply by Lilly-yang** [reply]
Hi,

Actually, there is no `map_server.launch` in the [mushr_sim](https://github.com/prl-mushr/mushr_sim/tree/master/launch) repository. I copied it from [mushr_base](https://github.com/prl-mushr/mushr_base/blob/master/mushr_base/launch/includes/map_server.launch) to my local `mushr_sim` directory. The map used is `sandbox.yaml`. Could you please send me the file(`$(find mushr_sim)/launch/map_server.launch`) in case it is different from the one in `mushr_base`.

Here is my O.S. information:
![Screenshot from 2022-04-08 18-37-01](https://user-images.githubusercontent.com/39558023/162485265-1e81c563-75fe-466e-90b6-6e3aaf13e5c7.png)

**Reply by sidtalia** [reply]
Hi @Lilly-yang,

Thank you for waiting, I was trying to reproduce this bug in a VM with the same OS and lower specs. 
I noticed that I was running an older version of mushr_sim and mushr_base, hence the earlier comment about using mushr_sim's map_server.launch. 

1) I then updated the mushr_sim, mush_base, and [vesc](https://github.com/prl-mushr/vesc.git) (if you already have vesc just make sure it is on master. I don't think this is the source of the problem though). **Remember to catkin_make after updating**

2) I ran the nhttc_ros nhttc_demo.launch as well as the multi_teleop.launch, and in both cases, the system ran without an issue (as in I got to the info messages that indicate that the node has started).

I assume that you're facing this bug even though your mushr_sim, mushr_base and nhttc_ros are up-to-date (on master). If that is the case and you're still facing issues, could you try running the following:
```
roslaunch mushr_sim single_car.launch
roslaunch mushr_base map_server.launch
```
and see if this creates an issue (it could be a different issue). Please also note the RAM usage when you launch these two nodes, I expect RAM usage to be around 1 GB. 

Regards,
Sidharth T.

**Reply by Lilly-yang** [reply]
Hi @naughtyStark ,

Today I tried again to launch `nhttc_ros/multi_teleop.launch`, and this time the 'node started" msg was successfully displayed.
![image](https://user-images.githubusercontent.com/39558023/162745038-2abd6b18-11ea-4f4b-9fdf-ad518bed0e6f.png)
I noticed that after I had tried launching many nodes. Even if no node is running, `multi_teleop.launch` gets stuck at 'log file: *.log' like before. In this case, I reboot the laptop and rerun it, and the nodes start again successfully. I don't know why.

After `multi_teleop.launch` is launched well(i.e. the node is started). I launched the `mushr_coordination` nodes. The good news is that `car1` moved. But the other three still didn't move. Their controllers died after the path was published. [recording screen](https://drive.google.com/file/d/1RgwYEoNQlQl71wcrr5tFo7NMGDuf_YLs/view?usp=sharing).
![image](https://user-images.githubusercontent.com/39558023/162745877-d8cc9dcd-58b3-4fe7-8d90-e174a07c8799.png)

I also tried launching these two nodes.
```
roslaunch mushr_sim single_car.launch
roslaunch mushr_base map_server.launch
```
only a warning popped up when launching `single_car.launch`
![image](https://user-images.githubusercontent.com/39558023/162746094-62164547-1faf-4daa-9398-9b6fe1cc9b5e.png)
and in rviz it showed up like this:
![Screenshot from 2022-04-11 14-21-48](https://user-images.githubusercontent.com/39558023/162746187-1c48115f-80fe-454c-9736-8407f90ea4d6.png)

I checked the RAM usage while I launched `nhttc_ros + mushr_coordination` and `mushr_sim + mushr_base`. It's about 4G/62.6G.

**Reply by sidtalia** [reply]
@Lilly-yang could you try running the commands from the multiagent navigation tutorial (the one that uses nhttc_demo.launch) and see if they work? I am not sure why the other agent nodes are crashing.

Regards,
Sidharth.

**Reply by Lilly-yang** [reply]
I have tried the `multi-agent navigation` tutorial before the `Multi-Agent Coordination` tutorial. It worked fine, except that the two cars didn't move simultaneously. You can check it out [here](https://drive.google.com/file/d/1LcxZI8b6LFRz6ijnsCcTJjXO-xR2tRtt/view?usp=sharing).

**Reply by sidtalia** [reply]
@Lilly-yang 
Hi, I am not sure why it isn't working on your system as expected. At this point, I would suggest running [this](https://drive.google.com/a/cs.washington.edu/file/d/1mOzSzVx9BF_e2U1OeK58NS42UIPcnIZq/view?usp=sharing) VM image from the [quickstart tutorial](https://mushr.io/tutorials/quickstart/) and then cloning the required repositories and seeing how that works out.

Please note that I may take a little longer to respond to queries. Our team is currently working on a docker version which should be out soon (can't say when exactly) and that may make things easier. Hopefully the VM approach works for you, given that you have ample computer power and RAM.

Regards,
Sidharth.
